src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -1601,21 +1597,31 @@ void Session::on_changesets_integrated(version_type client_version, const SyncPr
 {
     REALM_ASSERT_EX(m_state == Active, m_state);
     REALM_ASSERT_3(progress.download.server_version, >=, m_download_progress.server_version);
-    m_download_progress = progress.download;
     bool upload_progressed = (progress.upload.client_version > m_progress.upload.client_version);
+    bool download_progressed =
+        progress.download.server_version > m_progress.download.server_version &&
+        progress.download.last_integrated_client_version == m_progress.download.last_integrated_client_version;
```

Comment:
hmm. why this second check to see if the last_integrated_client_version hasn't progressed? i kinda thought this would be timing dependent on the server.

Reply:
Server version increases on every upload we make, and such DOWNLOAD message has no changesets in it.
Don't we hold off uploads until server signals that there are no more changesets for us to get?
Maybe i misunderstand this, but what is considered a progress in download?

Reply:
Tldr, The only thing that advances progress is receiving a DOWNLOAD, but there are no guarantees about what will be in the download that advances your progress.

We hold off on uploads when the session first starts, but after we receive the first batch of downloads to get caught up uploads can happen at any time. The server also will respond to an upload with an empty download if there are no further changesets to send but progress advanced, but the server could also just have more changesets to send and you'll do an upload and receive a download with changesets from before the upload and a new progress cursor indicating the server has accepted/persisted your upload.

Reply:
Replaced it to account for that i believe. These checks with client versions were not really needed.